### PR TITLE
[Fix uwp toolchain|world rebuild] make ninja work for uwp

### LINKED ISSUE #19601: [vcpkg|uwp] Fix toolchain so that ninja can be used
Copy default UWP build flags from MSBuild into the CMake toolchain. This removes the necessity to add extra flags in vcpkg_configure_(make|meson) and enables ninja builds with the uwp triplet.

### PR DESCRIPTION
**Pull request by Neumann-A**
~~depends #23001~~
~~depends #23846~~ (for less changes)
closes #19601

### CODE REVIEW

**Comment by JackBoosY** [comment]
This may fix #19818.

**Comment by Neumann-A** [comment]
> This may fix #19818.

be aware that the Fortran runtime libraries are not build against `WindowsApp.lib` which makes them not suiteable for UWP. So currently, Fortran cannot support UWP.

File: scripts/cmake/vcpkg_configure_make.cmake [hunk 1]
```diff
@@ -540,12 +541,15 @@ function(vcpkg_configure_make)
...
+            string(REPLACE "${_replacement}" "" VCPKG_DETECTED_CMAKE_CXX_FLAGS_RELEASE "${VCPKG_DETECTED_CMAKE_CXX_FLAGS_RELEASE}")
+            string(REPLACE "${_replacement}" "" VCPKG_DETECTED_CMAKE_C_FLAGS_RELEASE "${VCPKG_DETECTED_CMAKE_C_FLAGS_RELEASE}")
             # Can somebody please check if CMake's compiler flags for UWP are correct?
-            set(ENV{_CL_} "$ENV{_CL_} /D_UNICODE /DUNICODE /DWINAPI_FAMILY=WINAPI_FAMILY_APP /D__WRL_NO_DEFAULT_LIB_ -FU\"${VCToolsInstallDir}/lib/x86/store/references/platform.winmd\"")
-            string(APPEND VCPKG_DETECTED_CMAKE_CXX_FLAGS_RELEASE " -ZW:nostdlib")
-            string(APPEND VCPKG_DETECTED_CMAKE_CXX_FLAGS_DEBUG " -ZW:nostdlib")
-            set(ENV{_LINK_} "$ENV{_LINK_} ${VCPKG_DETECTED_CMAKE_C_STANDARD_LIBRARIES} ${VCPKG_DETECTED_CMAKE_CXX_STANDARD_LIBRARIES} /MANIFEST /DYNAMICBASE /WINMD:NO /APPCONTAINER") 
+            set(ENV{_CL_} "$ENV{_CL_} -FU\"${VCToolsInstallDir}/lib/x86/store/references/platform.winmd\"")
+            set(ENV{_LINK_} "$ENV{_LINK_} ${VCPKG_DETECTED_CMAKE_C_STANDARD_LIBRARIES} ${VCPKG_DETECTED_CMAKE_CXX_STANDARD_LIBRARIES}")
```
Comment (Neumann-A) [inline scripts/cmake/vcpkg_configure_make.cmake:552]: Theoretically this needs to be generalized to every flag containing spaces in its arguments.

**Comment by Neumann-A** [comment]
@BillyONeal: Please take a look at the uwp errors (except boost) and tell me which class of errors should be marked as `supports: "!uwp"`

**Comment by BillyONeal** [comment]
Unfortunately I know next to nothing about UWP; although if things built in the old msbuild world order it doesn't seem like a "shared" change like this should change which things are built successfully?

**Comment by Neumann-A** [comment]
> built in the old msbuild

As far as i know normal CI won't tell me if i passed a former regression and subsequent ports now fail.
Also cmake msbuild != vs default msbuild. Cmake is probably just using /ZW which is why it works. I'll compare some cmake msbuild build flags against this PR. Currently this PR uses just the vs defaults.

File: ports/vcpkg-cmake/vcpkg_cmake_configure.cmake [hunk 2]
```diff
@@ -282,6 +283,8 @@ function(vcpkg_cmake_configure)
     if(NOT DEFINED VCPKG_CHAINLOAD_TOOLCHAIN_FILE)
         if(VCPKG_TARGET_IS_MINGW)
             set(VCPKG_CHAINLOAD_TOOLCHAIN_FILE "${SCRIPTS}/toolchains/mingw.cmake")
+        elseif(VCPKG_TARGET_IS_UWP)
+            set(VCPKG_CHAINLOAD_TOOLCHAIN_FILE "${SCRIPTS}/toolchains/uwp.cmake")
```
Comment (Neumann-A) [inline ports/vcpkg-cmake/vcpkg_cmake_configure.cmake:287]: theoretically i could set this in the triplet instead of heare.

Comment (dg0yt) [reply ports/vcpkg-cmake/vcpkg_cmake_configure.cmake:287]: IMO this is something which should generally be done. But it probably needs a way to let the user override it.

Comment (Neumann-A) [reply ports/vcpkg-cmake/vcpkg_cmake_configure.cmake:287]: Overriding is always possible by defining your own triplet

File: ports/vcpkg-cmake/vcpkg_cmake_configure.cmake [hunk 3]
```diff
@@ -282,6 +283,8 @@ function(vcpkg_cmake_configure)
     if(NOT DEFINED VCPKG_CHAINLOAD_TOOLCHAIN_FILE)
         if(VCPKG_TARGET_IS_MINGW)
             set(VCPKG_CHAINLOAD_TOOLCHAIN_FILE "${SCRIPTS}/toolchains/mingw.cmake")
+        elseif(VCPKG_TARGET_IS_UWP)
```
Comment (ras0219-msft) [inline ports/vcpkg-cmake/vcpkg_cmake_configure.cmake:286]: This actually needs a tool change in https://github.com/microsoft/vcpkg-tool/blob/adfa9e24906d786e0e027064fac51b4e57f0c5ac/src/vcpkg/build.cpp#L816 because this affects binary caching. Specifically, the tool should define `VCPKG_CHAINLOAD_TOOLCHAIN_FILE` with its target toolchain if the triplet doesn't override it, which means we can get rid of all of the duplicate logic in this and other helpers -- they can always assume `VCPKG_CHAINLOAD_TOOLCHAIN_FILE` will be defined.

Comment (Neumann-A) [reply ports/vcpkg-cmake/vcpkg_cmake_configure.cmake:286]: > the tool should define

I would say the triplet should always define the toolchain file not the tool. The tool just needs to be informed if it should load vcvars or not and there is already a variable controlling that.
In that case you can get read of all of this big if/elseif blocks.

Comment (ras0219-msft) [reply ports/vcpkg-cmake/vcpkg_cmake_configure.cmake:286]: Opened https://github.com/microsoft/vcpkg-tool/pull/350 for that

**Comment by JackBoosY** [comment]
Since https://github.com/microsoft/vcpkg-tool/pull/350 was merged, can we continue this PR?

**Comment by Neumann-A** [comment]
still depends on https://github.com/microsoft/vcpkg/pull/23001 to move on.

File: ports/vcpkg-cmake/vcpkg_cmake_configure.cmake (hunk 3 above)
Comment (Neumann-A) [reply ports/vcpkg-cmake/vcpkg_cmake_configure.cmake:286]: https://github.com/microsoft/vcpkg/pull/23846 will use that change.

**Comment by Neumann-A** [comment]
@JackBoosY Found the error: The ports failing are all using MSBuild since `PREFER_NINJA` is missing. For some reason CMake + VS Generator chokes on `/WINMD` in the  `CMAKE_EXE_LINKER_FLAGS`. I also observed that the compiler check is not correctly using the `/FU` flag (containing and ugly space) so i had to move it into the environment variable.

**Comment by Neumann-A** [comment]
```
REGRESSION: physfs:arm-uwp # CMake Error: TRY_RUN() invoked in cross-compiling mode, please set the following cache variables appropriately:    HAS_PRE_1970_EXITCODE (advanced)   HAS_PRE_1970_EXITCODE__TRYRUN_OUTPUT (advanced)
REGRESSION: orc:arm-uwp # D:\buildtrees\physfs\src\3.0.2-5516c52307.clean\src\physfs_platform_winrt.cpp(29): error C2653: 'Windows': is not a class or namespace name
REGRESSION: physx:arm-uwp # D:\buildtrees\physx\src\be49a99479-2db4990732.clean\pxshared\include\foundation/PxPreprocessor.h(449): fatal error C1189: #error:  Exactly one of NDEBUG and _DEBUG needs to be defined!
```

so physfs is not setup for cross builds -> probably ci baseline it
orc seems to have simple syntax errors (maybe missing include) -> upstream issue -> ci baseline?
physx needs /DNDEBUG in release builds -> fix by adding `/DNDEBUG` to the toolchains `_RELEASE` compiler flags

**Comment by Neumann-A** [comment]
Yeah uwp is green

**Comment by JackBoosY** [comment]
```
Errors occurred while parsing C:\a\1\s\ports\igraph\vcpkg.json
    $.port-version: mismatched type: expected a nonnegative integer
```

File: ports/orc/portfile.cmake [hunk 4]
```diff
@@ -16,21 +16,21 @@ file(REMOVE "${SOURCE_PATH}/cmake_modules/FindProtobuf.cmake")
...
   set(BUILD_TOOLS OFF)
 else()
   set(BUILD_TOOLS ON)
 endif()
 
+if(NOT VCPKG_TARGET_IS_UWP)
+    set(configure_opts PREFER_NINJA)
+endif()
+
 vcpkg_configure_cmake(
```
Comment (Thomas1664) [inline ports/orc/portfile.cmake:31]:
```suggestion
vcpkg_cmake_configure(
```
Has the port a problem with the new one?

File: ports/physx/vcpkg.json [hunk 5]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "physx",
-  "version-semver": "4.1.2",
-  "port-version": 1,
+  "version": "4.1.2",
+  "port-version": 3,
```
Comment (Thomas1664) [inline ports/physx/vcpkg.json:4]:
```suggestion
  "port-version": 2,
```
Or is master already at 2?

**Review by Thomas1664 — CHANGES_REQUESTED** [review]

File: ports/physx/vcpkg.json (hunk 5 above)
Comment (Neumann-A) [reply ports/physx/vcpkg.json:4]: Could be that there is a 2 in the version history for some strange reason.

Comment (Thomas1664) [reply ports/physx/vcpkg.json:4]: > Could be that there is a 2 in the version history for some strange reason.

I checked it on master: This should really be 2. The latest `port-version` is 1

File: ports/orc/portfile.cmake (hunk 4 above)
Comment (Thomas1664) [reply ports/orc/portfile.cmake:31]: And this should really be `vcpkg_cmake_configure` since you removed `PREFER_NINJA`

File: ports/physx/vcpkg.json (hunk 5 above)
Comment (Neumann-A) [reply ports/physx/vcpkg.json:4]: > I checked it on master: This should really be 2. The latest port-version is 1

did you check the version db? I know in one of the ports was a strange entry where 2 already existed but the port-version was 1

File: ports/orc/portfile.cmake (hunk 4 above)
Comment (Neumann-A) [reply ports/orc/portfile.cmake:31]: I didn't remove it.... It is just not added for UWP. If you want this changed please open up a PR against my fork so I can merge it. I really don't have time to update all those small details.

File: ports/physx/vcpkg.json (hunk 5 above)
Comment (Thomas1664) [reply ports/physx/vcpkg.json:4]: I checked `versions\p-\physx.json` and `versions\baseline.json`. Both are on `"port-version": 1`.

File: ports/orc/portfile.cmake (hunk 4 above)
Comment (Thomas1664) [reply ports/orc/portfile.cmake:31]: Have a look at 2 lines below from the files view. The diff shows that you replaced `PREFER_NINJA` with `${configure_opts}`.  But I can open a PR anyway.

File: ports/physx/vcpkg.json (hunk 5 above)
Comment (Neumann-A) [reply ports/physx/vcpkg.json]: you broke orc please fix it..... Also don't use `version-semver`

File: ports/sdl2/portfile.cmake [hunk 6]
```diff
@@ -32,8 +32,13 @@ if ("x11" IN_LIST FEATURES)
     message(WARNING "You will need to install Xorg dependencies to use feature x11:\nsudo apt install libx11-dev libxft-dev libxext-dev\n")
 endif()
 
+if(VCPKG_TARGET_IS_UWP)
+    set(configure_opts WINDOWS_USE_MSBUILD)
+endif()
```
Comment (JackBoosY) [inline ports/sdl2/portfile.cmake:37]: So why do we need to use msbuild in uwp? This PR should fix this.

Comment (Neumann-A) [reply ports/sdl2/portfile.cmake:37]: there are always exceptions from the rule. There are also CMake properties which only work for certain generators. 
You can try to remove this and try to debug it but I won't be spending that time, since this is already the current baseline  behavior. 

If 95% of the ports now work with ninja and only 5 % not that is already a big improvement to the current state. Furthermore I think that this PR enables a lot of ports for UWP.

File: ports/igraph/vcpkg.json [hunk 7]
```diff
@@ -1,8 +1,10 @@
 {
   "name": "igraph",
   "version": "0.9.7",
+  "port-version": 1,
   "description": "igraph is a C library for network analysis and graph theory, with an emphasis on efficiency portability and ease of use.",
   "homepage": "https://igraph.org/",
+  "supports": "!uwp",
```
Comment (JackBoosY) [inline ports/igraph/vcpkg.json:7]: This dependency metis failed to build previously on uwp.

**Review by JackBoosY — APPROVED** [review]

File: ports/igraph/vcpkg.json (hunk 7 above)
Comment (Neumann-A) [reply ports/igraph/vcpkg.json:7]: I checked the sources for this and it does not support uwp. (Using LoadLibrary if I remember correctly)

**Comment by JackBoosY** [comment]
So I think my PR https://github.com/microsoft/vcpkg/pull/21537 can be closed?

File: ports/blas/CMakeLists.txt [hunk 8]
```diff
@@ -1,4 +1,4 @@
 cmake_minimum_required(VERSION 3.17)
 
-project(Find_BLAS_external)
+project(Find_BLAS_external LANGUAGES CXX)
```
Comment (Neumann-A) [inline ports/blas/CMakeLists.txt:3]:
```suggestion
project(Find_BLAS_external)
```

unnecessary change. Will be removed after team review.

File: ports/blas/vcpkg.json [hunk 9]
```diff
@@ -1,9 +1,12 @@
 {
   "name": "blas",
-  "version-string": "1",
-  "port-version": 1,
+  "version-date": "2022-03-31",
   "description": "Metapackage for packages which provide BLAS",
```
Comment (Neumann-A) [inline ports/blas/vcpkg.json:4]:
```suggestion
  "description": "Metapackage for packages which provide BLAS",
  "license": null,
```
Unimportant change. Will be added after team review.

File: ports/ffmpeg/portfile.cmake [hunk 10]
```diff
@@ -112,9 +112,9 @@ if(VCPKG_DETECTED_MSVC)
     set(OPTIONS "--toolchain=msvc ${OPTIONS}")
     # This is required because ffmpeg depends upon optimizations to link correctly
     string(APPEND VCPKG_COMBINED_C_FLAGS_DEBUG " -O2")
-    string(REPLACE " -RTC1 " "" VCPKG_COMBINED_C_FLAGS_DEBUG "${VCPKG_COMBINED_C_FLAGS_DEBUG}")
-    string(REPLACE " -Od " "" VCPKG_COMBINED_C_FLAGS_DEBUG "${VCPKG_COMBINED_C_FLAGS_DEBUG}")
-    string(REPLACE " -Ob0 " "" VCPKG_COMBINED_C_FLAGS_DEBUG "${VCPKG_COMBINED_C_FLAGS_DEBUG}")
+    string(REPLACE " -RTC1" "" VCPKG_COMBINED_C_FLAGS_DEBUG "${VCPKG_COMBINED_C_FLAGS_DEBUG}")
+    string(REPLACE " -Od" "" VCPKG_COMBINED_C_FLAGS_DEBUG "${VCPKG_COMBINED_C_FLAGS_DEBUG}")
+    string(REPLACE " -Ob0" "" VCPKG_COMBINED_C_FLAGS_DEBUG "${VCPKG_COMBINED_C_FLAGS_DEBUG}")
```
Comment (Neumann-A) [inline ports/ffmpeg/portfile.cmake:117]: This one could be made into a separate PR. Has nothing to do with this PR just fixes an observed build error with the introduced changes.

File: ports/vcpkg-cmake/vcpkg_cmake_configure.cmake [hunk 11]
```diff
@@ -156,7 +156,8 @@ function(vcpkg_cmake_configure)
         set(ninja_host OFF)
     elseif(VCPKG_TARGET_IS_UWP)
         # Ninja and MSBuild have many differences when targetting UWP, so use MSBuild to maximize existing compatibility
```
Comment (Neumann-A) [inline ports/vcpkg-cmake/vcpkg_cmake_configure.cmake:158]:
```suggestion
```
Removal. Will be added after team review.

File: ports/ffmpeg/portfile.cmake (hunk 10 above)
Comment (JackBoosY) [reply ports/ffmpeg/portfile.cmake:117]: I think it's good to me. You can leave this changes in this PR.

File: ports/vcpkg-cmake/vcpkg_cmake_configure.cmake [hunk 12]
```diff
@@ -156,7 +156,8 @@ function(vcpkg_cmake_configure)
...
-        set(ninja_can_be_used OFF)
+        # set(ninja_can_be_used OFF)
+        # Nonsense windows.cmake does just not correctly setup the compiler flags for UWP and Ninja
```
Comment (Neumann-A) [inline ports/vcpkg-cmake/vcpkg_cmake_configure.cmake:160]:
```suggestion
```
Removal. Will be added after team review.

File: scripts/cmake/vcpkg_build_make.cmake [hunk 13]
```diff
@@ -148,7 +148,7 @@ function(vcpkg_build_make)
                 set(LINKER_FLAGS_${cmake_buildtype} "${VCPKG_DETECTED_CMAKE_STATIC_LINKER_FLAGS_${cmake_buildtype}}")
             else() # dynamic
                 set(LINKER_FLAGS_${cmake_buildtype} "${VCPKG_DETECTED_CMAKE_SHARED_LINKER_FLAGS_${cmake_buildtype}}")
-            endif()            
+            endif()
```
Comment (Neumann-A) [inline scripts/cmake/vcpkg_build_make.cmake:151]:
```suggestion
            endif()            
```

WS change. Will be removed after team review

File: scripts/cmake/vcpkg_configure_cmake.cmake [hunk 14]
```diff
@@ -155,7 +155,8 @@ function(vcpkg_configure_cmake)
             set(ninja_host OFF)
         elseif(VCPKG_TARGET_IS_UWP)
             # Ninja and MSBuild have many differences when targeting UWP, so use MSBuild to maximize existing compatibility
-            set(ninja_can_be_used OFF)
+            # set(ninja_can_be_used OFF)
+            # Nonsense windows.cmake does just not correctly setup the compiler flags for UWP and Ninja
```
Comment (Neumann-A) [inline scripts/cmake/vcpkg_configure_cmake.cmake:159]: Remove after review
```suggestion
```

File: scripts/cmake/vcpkg_configure_meson.cmake [hunk 15]
```diff
@@ -205,7 +205,11 @@ function(z_vcpkg_meson_generate_native_file_config config_type) #https://mesonbu
 
     if(NOT VCPKG_CHAINLOAD_TOOLCHAIN_FILE)
         if(VCPKG_TARGET_IS_WINDOWS AND NOT VCPKG_TARGET_IS_MINGW)
-            set(VCPKG_CHAINLOAD_TOOLCHAIN_FILE "${SCRIPTS}/toolchains/windows.cmake")
+            if(VCPKG_TARGET_IS_UWP)
+                set(VCPKG_CHAINLOAD_TOOLCHAIN_FILE "${SCRIPTS}/toolchains/uwp.cmake")
+            else()
+                set(VCPKG_CHAINLOAD_TOOLCHAIN_FILE "${SCRIPTS}/toolchains/windows.cmake")
+            endif()
```
Comment (Neumann-A) [inline scripts/cmake/vcpkg_configure_meson.cmake:212]:
```suggestion
            set(VCPKG_CHAINLOAD_TOOLCHAIN_FILE "${SCRIPTS}/toolchains/windows.cmake")
```
revert after https://github.com/microsoft/vcpkg/pull/23846 is merged

**Comment by Neumann-A** [comment]
> So I think my PR https://github.com/microsoft/vcpkg/pull/21537 can be closed?

Probably ;).

File: ports/ffmpeg/portfile.cmake (hunk 10 above)
Comment (Neumann-A) [reply ports/ffmpeg/portfile.cmake:117]: >  You can leave this changes in this PR.

I didn't mean to remove it. it is just that this change is fixing a port-bug which is not the main topic of this PR.

Comment (Neumann-A) [reply ports/ffmpeg/portfile.cmake:117]: I opened up #24077 because I think this here is incomplete and needs separate review.

File: scripts/get_cmake_vars/CMakeLists.txt [hunk 16]
```diff
@@ -92,12 +92,16 @@ endforeach()
 macro(_vcpkg_adjust_flags flag_var)
     if(MSVC) # Transform MSVC /flags to -flags due to bash scripts intepreting /flag as a path.
         string(REGEX REPLACE "(^| )/" "\\1-" ${flag_var} "${${flag_var}}")
+        if(CMAKE_SYSTEM_NAME STREQUAL "WindowsStore")
+            foreach("${flag_var}" IN_LIST CMAKE_C_FLAGS CMAKE_CXX_FLAGS)
```
Comment (strega-nil) [inline scripts/get_cmake_vars/CMakeLists.txt:96]:
```suggestion
            foreach("${flag_var}" IN_LIST CMAKE_C_FLAGS CMAKE_CXX_FLAGS)
```

this is a very, very weird line; what this will do is set `${flag_var}` to first `IN_LIST` (and append "-ZW:nostdlib" to `IN_LIST`), then the same for CMAKE_C_FLAGS and CMAKE_CXX_FLAGS; did you mean to do something like

```suggestion
            foreach(var IN ITEMS CMAKE_C_FLAGS CMAKE_CXX_FLAGS)
```
?

**Review by strega-nil — CHANGES_REQUESTED** [review]
I mean I love the idea, but I'm definitely not an expert; you'll have to poke Robert to get a review

File: scripts/get_cmake_vars/CMakeLists.txt (hunk 16 above)
Comment (Neumann-A) [reply scripts/get_cmake_vars/CMakeLists.txt:96]: Good catch. It should simply be a `if()` block not a `foreach()` block. Maybe a merge messed it up.

**Comment by Neumann-A** [comment]
> I mean I love the idea, but I'm definitely not an expert; you'll have to poke Robert to get a review

cc @ras0219-msft

File: ports/vcpkg-cmake/cmake_get_vars/CMakeLists.txt [hunk 17]
```diff
@@ -147,6 +147,11 @@ foreach(flag CXX C SHARED_LINKER EXE_LINKER STATIC_LINKER RC)
         # Transform MSVC /flags to -flags due to bash scripts intepreting /flag as a path.
         # This is imperfect because it fails on directories with trailing spaces, but those are exceedingly rare
         string(REGEX REPLACE "(^| )/" "\\1-" ${flag}_FLAGS "${${flag}_FLAGS}")
+        if(CMAKE_SYSTEM_NAME STREQUAL "WindowsStore")
+            if("${flag}" MATCHES "^(C|CXX)$")
```
Comment (Neumann-A) [inline ports/vcpkg-cmake/cmake_get_vars/CMakeLists.txt:151]:
```suggestion
            if("${flag}" STREQUAL "CXX")
```

File: scripts/get_cmake_vars/CMakeLists.txt [hunk 18]
```diff
@@ -92,12 +92,16 @@ endforeach()
 macro(_vcpkg_adjust_flags flag_var)
     if(MSVC) # Transform MSVC /flags to -flags due to bash scripts intepreting /flag as a path.
         string(REGEX REPLACE "(^| )/" "\\1-" ${flag_var} "${${flag_var}}")
+        if(CMAKE_SYSTEM_NAME STREQUAL "WindowsStore")
+            if("${flag_var}" MATCHES "^(CMAKE_C_FLAGS|CMAKE_CXX_FLAGS)$")
```
Comment (Neumann-A) [inline scripts/get_cmake_vars/CMakeLists.txt:96]:
```suggestion
            if("${flag_var}" STREQUAL "CMAKE_CXX_FLAGS")
```

**Comment by BillyONeal** [comment]
Thanks for making this work @Neumann-A : Words can't describe how much I appreciate the pain you went through to get here trying to demystify the UWP black box. I'm going to try to confirm with some folks who own what UWP means tomorrow.

**Comment by BillyONeal** [comment]
image

**Comment by BillyONeal** [comment]
image

File: scripts/toolchains/uwp.cmake [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+
+    if(VCPKG_CRT_LINKAGE STREQUAL "dynamic")
+        set(VCPKG_CRT_LINK_FLAG_PREFIX "/MD")
+    elseif(VCPKG_CRT_LINKAGE STREQUAL "static")
+        set(VCPKG_CRT_LINK_FLAG_PREFIX "/MT")
+    else()
+        message(FATAL_ERROR "Invalid setting for VCPKG_CRT_LINKAGE: \"${VCPKG_CRT_LINKAGE}\". It must be \"static\" or \"dynamic\"")
+    endif()
+
+    set(_vcpkg_charset "/utf-8")
```
Comment (BillyONeal) [inline scripts/toolchains/uwp.cmake:16]: I'm a bit sad that this one is called `_vcpkg_charset` but the equivalent in the Windows toolchain is `CHARSET_FLAG`. No change requested (unless you want).

File: scripts/toolchains/uwp.cmake [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+    if (NOT VCPKG_SET_CHARSET_FLAG OR VCPKG_PLATFORM_TOOLSET MATCHES "v120")
+        # VS 2013 does not support /utf-8
+        set(_vcpkg_charset "")
+    endif()
+
+    set(_vcpkg_cpp_flags "/DWIN32 /D_WINDOWS /D_UNICODE /DUNICODE /DWINAPI_FAMILY=WINAPI_FAMILY_APP /D__WRL_NO_DEFAULT_LIB__" ) # VS adds /D "_WINDLL" for DLLs;
+    set(_vcpkg_common_flags "/nologo /Z7 /MP /GS /Gd /Gm- /W3 /WX- /Zc:wchar_t /Zc:inline /Zc:forScope /fp:precise /Oy- /EHsc")
```
Comment (BillyONeal) [inline scripts/toolchains/uwp.cmake:23]:
```suggestion
    set(_vcpkg_common_flags "/nologo /Z7 /GS /Gd /Gm- /W3 /WX- /Zc:wchar_t /Zc:inline /Zc:forScope /fp:precise /Oy- /EHsc") 
```

`/MP` is a workaround for an MSBuildism that doesn't apply here.

**Review by BillyONeal — APPROVED** [review]
I am giving folks who understand UWP better a few days to confirm that this is reasonable, but this is awesome.

THANK YOU SO MUCH!

File: scripts/toolchains/uwp.cmake (hunk 20 above)
Comment (BillyONeal) [reply scripts/toolchains/uwp.cmake:23]: (Don't worry about applying this yet; I'll just apply it before merging. Just avoiding building the world again until I'm ready to push the merge button)

**Comment by BillyONeal** [comment]
@JackBoosY I'm waiting for internal folks to OK it, @Neumann-A doesn't need to respond unless they want. :)

File: scripts/toolchains/uwp.cmake (hunk 19 above)
Comment (Neumann-A) [reply scripts/toolchains/uwp.cmake:16]: you can change this with the` /MP` change below however you like. I just mimicked the windows toolchain.

Comment (BillyONeal) [reply scripts/toolchains/uwp.cmake:16]: The flag is pointless, but isn't worth another world rebuild to change if that's the only remaining nitpick. You got it to be green, so I'm going to run with it :)

**Comment by BillyONeal** [comment]
Thanks again for getting UWP to not be a special snowflake!
